app/src/components/modal/modal_deposit_withdraw/index.tsx

```diff
@@ -103,224 +101,200 @@ interface Props extends HTMLAttributes<HTMLDivElement> {
-    }
+  const findCurrentTokenBasedOnAction = (exchange: ExchangeType, symbol: string): Token | undefined => {
+    if (exchange === ExchangeType.deposit) {
+      return (
+        mainnetTokens.find(
+          token => token.symbol === (context.networkId === networkIds.XDAI && symbol === 'DAI' ? 'xDAI' : symbol),
+        ) || mainnetTokens.find(token => token.symbol === symbol)
+      )
+    } else if (exchange === ExchangeType.withdraw && symbol === 'DAI') {
+      const nativeAsset = getNativeAsset(networkIds.XDAI)
+
+      return { ...nativeAsset, balance: xDaiBalance ? xDaiBalance.toString() : '0' }
+    } else return xDaiTokens.find(token => token.symbol === (symbol === 'DAI' ? 'xDAI' : symbol))
   }
+  const { address, balance, decimals, symbol }: any = findCurrentTokenBasedOnAction(
+    exchangeType,
+    currencySelected.toUpperCase(),
+  )
+  const currentTokenMainnet = mainnetTokens.find(element => element.symbol === symbol)
+
+  const isApprovalVisible =
+    (exchangeType === ExchangeType.deposit &&
+      currentTokenMainnet &&
+      new BigNumber(currentTokenMainnet.allowance ? currentTokenMainnet.allowance : '0').isZero() &&
+      displayFundAmount.gte(currentTokenMainnet.allowance || Zero)) ||
+    (exchangeType === ExchangeType.deposit &&
+      initiatedAllowanceState === ButtonStates.finished &&
+      currentTokenMainnet &&
+      !new BigNumber(currentTokenMainnet.allowance ? currentTokenMainnet.allowance : '0').isZero())
 
   const approve = async () => {
     try {
+      setInitiatedAllowanceState(ButtonStates.working)
       if (exchangeType === ExchangeType.deposit) {
-        if (currencySelected === ExchangeCurrency.Omen) {
-          setOmenAllowanceState(ButtonStates.working)
-          const collateralService = new ERC20Service(context.rawWeb3Context.library, account, omenToken.address)
+        if (currencySelected === 'dai') {
+          const collateralService = new ERC20Service(context.rawWeb3Context.library, account, address)
 
-          await collateralService.approveUnlimited(OMNI_BRIDGE_MAINNET_ADDRESS)
-          setOmenAllowanceState(ButtonStates.finished)
+          await collateralService.approveUnlimited(DAI_TO_XDAI_TOKEN_BRIDGE_ADDRESS)
         } else {
-          setDaiAllowanceState(ButtonStates.working)
-          const collateralService = new ERC20Service(context.rawWeb3Context.library, account, DAI.address)
+          const collateralService = new ERC20Service(context.rawWeb3Context.library, account, address)
 
-          await collateralService.approveUnlimited(DAI_TO_XDAI_TOKEN_BRIDGE_ADDRESS)
-          setDaiAllowanceState(ButtonStates.finished)
+          await collateralService.approveUnlimited(OMNI_BRIDGE_MAINNET_ADDRESS)
         }
       }
-
-      await fetchAllowance()
-    } catch (e) {
-      if (currencySelected === ExchangeCurrency.Omen) setOmenAllowanceState(ButtonStates.idle)
-      else setDaiAllowanceState(ButtonStates.idle)
+      await fetchBalances()
```

Comment:
`fetchBalances()` is not of type `<Promise>` so i don't believe the `await` key word is necessary here

Reply:
in 99% of cases `fetchBalances()` comes from `  const { fetchBalances } = useConnectedBalanceContext()` in which case we do use the `await` keyword. in your PR `fetchBalances` is the `props`, so, is it necessary to use `await`? i am not entirely certain but it was the only thing I could find sleuthing through the code.

Reply:
I think just the `fetchBalances` type is incorrect, it actually is async - it don't matter though I will mess with this stuff later in https://github.com/protofire/omen-exchange/pull/1934 once this PR is merged

Reply:
@hexyls Yeah I just looked at it type is wrong i will change it to. fetchBalances: () => Promise<void>